Add praw.models.LiveHelper.__call__

nmtake opened the pull request:
## Feature Summary and Justification

This feature provides explicit interface to initialize a `praw.models.LiveThread` instance, e.g.,

```python
>>> import praw
>>> from  praw.models import LiveThread
>>> reddit = praw.Reddit(...)
>>> thread = LiveThread(reddit, 'ukaeu1ik4sw5')
>>> thread.title
u'reddit updates'
```

## References

* nmtake/praw@e147a37

## Concerns

Access tokens are visible in the cassettes. I think it would be better if we could filter out those tokens with `before_record` hook.  Any thoughts?

bboe commented on CHANGES.rst:
```diff
@@ -8,6 +8,8 @@ Unreleased
 
 * :meth:`praw.models.Subreddits.search_by_topic` to search subreddits by topic.
   (see: https://www.reddit.com/dev/api/#GET_api_subreddits_by_topic).
+* :meth:`praw.models.LiveThread.__init__` to provide explicit interface for
```
Can we add the interface to this feature at this time as I want to discourage directly creating instances of the classes. The way I think we talked about creating instance of live threads was via:

    reddit.live(thread_id)

Then the changelog can link to this method rather than an `__init__` method.

With this it would be great to add a unit test similar to `test_comment` for this feature to ensure the LiveThread is lazy: https://github.com/praw-dev/praw/blob/de1caa50deed07f19c2530f7894e575ec1b88ea9/tests/unit/test_reddit.py#L29

nmtake replied:
> Can we add the interface to this feature at this time as I want to discourage directly creating instances of the classes. 

Sure! Should I change this PR title according to your suggestion in this case? I don't get used to github and git yet.

> With this it would be great to add a unit test similar to test_comment for this feature to ensure the LiveThread is lazy

Okay. Thanks for the reference.

bboe commented on tests/integration/cassettes/TestLiveThread_test_init.json:
```diff
@@ -0,0 +1,164 @@
+        "headers": {
+          "Accept": "*/*",
+          "Accept-Encoding": "identity",
+          "Authorization": "Basic <BASIC_AUTH>",
```
Did you have any issues or have to find a reference to be able to set the environment variables to get these secrets removed?

nmtake replied:
I've read conftest.py so no issues for me. Anyway, I think [this comment](https://github.com/praw-dev/praw/pull/679#pullrequestreview-10550244) is really useful for contributors.

bboe replied:
Yes, I should add that to the contributors file and make sure that the file is part of the documentation (it currently is not).

bboe commented on tests/integration/cassettes/TestLiveThread_test_init.json:
```diff
@@ -0,0 +1,164 @@
+      "response": {
+        "body": {
+          "encoding": "UTF-8",
+          "string": "{\"access_token\": \"3ECWseFgsClB1rG7H0GMLSnHI9A\", \"token_type\": \"bearer\", \"expires_in\": 3600, \"scope\": \"*\"}"
```
Since the access tokens are only valid for an hour I personally am not terribly worried about them. Also since I also only generate access tokens for the pyapitestuser2 account.

However, if you want to make a separate PR that filters out these access tokens transparently, I think that would be awesome.

nmtake replied:
Yeah I believe most contributors will use dedicated account for testing, or revoke the token after testing, but I think it would be better if some safeguard exists.

> However, if you want to make a separate PR that filters out these access tokens transparently, I think that would be awesome.

Okay, I'll try to do that.

bboe commented on tests/integration/models/reddit/test_live_thread.py:
```diff
@@ -0,0 +1,26 @@
+            assert thread.id == thread_id
+            assert thread.title == 'reddit updates'
+
+            with pytest.raises(TypeError):
```
Can you add a unit test for this part since it doesn't require any network connectivity?

It would be nice for the LiveThread unit test to contain many of the similar tests as in https://github.com/praw-dev/praw/blob/de1caa50deed07f19c2530f7894e575ec1b88ea9/tests/unit/models/reddit/test_comment.py

nmtake replied:
Sure. Thanks for the link again.

bboe commented:
Please ping me when you've updated as github doesn't send notifications when the commits for a PR have been updated.

nmtake commented:
> Please ping me when you've updated as github doesn't send notifications when the commits for a PR have been updated.

Okay.

nmtake commented:
Updated. Now the PR implements `praw.models.LiveHelper.__call__()` and `praw.models.LiveThread.__init__()`. 
```
>>> import praw
>>> reddit = praw.Reddit(...)
>>> thread = reddit.live("ukaeu1ik4sw5")
>>> thread.title
u'reddit updates'
```

I've added unit tests for LiveThread, but I have some concerns:

1. I've overrided `__eq__` so that LiveThread IDs are case sensitive (see [unit test](https://github.com/praw-dev/praw/pull/680/commits/d8f29d7bd764db305a2d3e2571351bdb6a45ce03#diff-5444a4e97c1ccf958f3788925259e7dbR21)).
2. I'm not sure whether `_fetched` flag should be set or not when LiveThread is instantiated with ID, so related lines ([this](https://github.com/praw-dev/praw/pull/680/commits/d8f29d7bd764db305a2d3e2571351bdb6a45ce03#diff-ddfe088db1ec17b1133e057c48bd948cR34) and [this](https://github.com/praw-dev/praw/pull/680/commits/d8f29d7bd764db305a2d3e2571351bdb6a45ce03#diff-5444a4e97c1ccf958f3788925259e7dbR60) are commented out.

bboe commented on praw/models/reddit/live.py:
```diff
@@ -8,5 +8,31 @@ class LiveThread(RedditBase):
+        super(LiveThread, self).__init__(reddit, _data)
+        if id:
+            self.id = id  # pylint: disable=invalid-name
+        # else:
```
The comment here can be deleted since at the moment there is no where in PRAW's code that is explicitly passing the data attribute -- that's unique to comments.

nmtake replied:
Hmm, I couldn't find the code like `Comment(reddit, _data=...)` but with some grep I found  Reddior class is instantiated with `_data` in [models/user.py](https://github.com/praw-dev/praw/blob/3e1bb8a26d6c5251dc5b7976591e05bc9c690d67/praw/models/user.py#L61):

```python
    def me(self, use_cache=True):  # pylint: disable=invalid-name
        ...
        if self._me is None or not use_cache:
            user_data = self._reddit.get(API_PATH['me'])
            self._me = Redditor(self._reddit, _data=user_data)
return self._me
```

bboe replied:
Good sleuthing -- I was wrong. The comments I was referring to are auto populated as part of obtaining a Submission, but weren't considered fetched at that point even though they had all the data they needed.

bboe commented on praw/models/reddit/live.py:
```diff
@@ -8,5 +8,31 @@ class LiveThread(RedditBase):
     STR_FIELD = 'id'
 
+    def __eq__(self, other):
+        """Return whther the other instance equals the current."""
```
whther -> whether

Might be worth noting here and in `__hash__` that this comparison is case sensitive whereas the super class's comparisons are not.

nmtake replied:
Yes, it should be noted. I'll do that in the updated PR.

bboe commented on praw/models/helpers.py:
```diff
@@ -3,12 +3,20 @@
+    def __call__(self, id):  # pylint: disable=invalid-name,redefined-builtin
+        """Return a new lazy instance of :class:`~.LiveThread`.
+
+        :param id: A live thread ID, e.g., ``ukaeu1ik4sw5``.
```
I know the other helpers don't yet have it, but could we provide an example of usage here so that people don't try to do `LiveHelper(reddit).__call__(id)`? This is something I've seen.

Adding something like the following should go a long way to make these docstrings, which are generated into RTD significantly more usable:

```rst

This method is intended to be used as:

.. code:: python
   livethread = reddit.live('ukaeu1ik4sw5')
```

nmtake replied:
It's great to include a usage in a docstring. I'll try that.

bboe commented on tests/integration/cassettes/TestLiveThread_test_init.json:
```diff
@@ -0,0 +1,164 @@
+      "response": {
+        "body": {
+          "encoding": "UTF-8",
+          "string": "{\"access_token\": \"tuIhJlqr1dLVDqCFbVqtWF7tynw\", \"token_type\": \"bearer\", \"expires_in\": 3600, \"scope\": \"*\"}"
```
I was thinking about the access token again. Given that this access token requires both the client id and client secret, it's effectively unusable without those even within the 1-hour timeframe.

Of course, I'm still open to a PR that removes the access token just to save people from the thought that they might be able to use the tokens.

nmtake replied:
No, reddit's OAuth2 access token is solely usable without client_id and secret. Please see the interactions in the cassete and https://github.com/reddit/reddit/wiki/OAuth2-Quick-Start-Example#python-example

bboe replied:
Wow, despite writing the OAuth code twice, I mistakenly assumed that the authorization header was also sent in each request. Oops. Thanks for the correction.

There certainly is reason to try to hide the tokens in such cases. Thanks for the correction.

bboe commented on tests/integration/models/reddit/test_live.py:
```diff
@@ -0,0 +1,21 @@
+        data = {'id': thread_id}
+
+        with self.recorder.use_cassette('TestLiveThread_test_init'):
+            thread = LiveThread(self.reddit, thread_id)
```
If this is lazy we should probably put this line and the next outside the betamax block.

bboe commented on tests/integration/test_reddit.py:
```diff
@@ -23,6 +23,14 @@ def test_info(self):
+    def test_live_call(self, _):
+        thread_id = 'ukaeu1ik4sw5'
+        with self.recorder.use_cassette('TestReddit.test_live_call'):
+            thread = self.reddit.live(thread_id)
```
This line in the next seem like they need not require a network request.

bboe commented on tests/unit/models/reddit/test_live.py:
```diff
@@ -0,0 +1,64 @@
+        assert thread1 == thread1
+        assert thread2 == thread2
+        assert thread3 == thread3
+        assert thread1 != thread2  # live thread ID in a URL is case sensitive
```
⭐️ great realization that live thread ids are case sensitive. I definitely would have missed that.

nmtake replied:
Yes, at least https://www.reddit.com/live/UKaeu1ik4sw5/ (note thread id's first 2 characters are upper cased) returns 404.

bboe commented on tests/unit/models/reddit/test_live.py:
```diff
@@ -0,0 +1,64 @@
+        thread = LiveThread(self.reddit, id='dummy')
+        assert str(thread) == 'dummy'
+
+#    def test_unset_hidden_attribute_does_not_fetch(self):
```
This can be removed as it was a special case for comments where printing a comment in ipython resulted in a fetch that removed its replies.

nmtake commented:
Updated according to your suggestions. Also, I've moved some test code which requires no network access from integration test to [unit test](https://github.com/praw-dev/praw/pull/680/commits/8ea35c50f5657cb7be8cfd140b8495a46d7c6553#diff-5444a4e97c1ccf958f3788925259e7dbR11).

bboe commented:
Really great work! Thanks for the contribution and the discussion. I am very much looking forward to your future contributions. 🎉

nmtake commented:
Thanks for merging!  I'll continue to implement the rest of reddit live features.
